Write introduction

tmravic opened the pull request:
<!--
👋 Hi, thanks for sending a PR to Codecademy Docs! 💖
Please fill out all fields below to ensure your PR is reviewed quickly.

**IMPORTANT**

If you would like to receive credit for your contribution to an entry, make sure to have your Codecademy user profile linked to your GitHub account:

1. Go to your Codecademy dashboard.
2. Click on your profile image in the top-right and then choose "Profile".
3. Then click "Edit Profile".
4. In the "GitHub Username" field, add your username (without the @).
5. Click "Save Changes"

Of course, you can opt not to do this and be listed as an "Anonymous contributor", instead. :)
-->

### Description

This PR add a description of the usage of common methods of ARGV and ARGF in Ruby for reading files from Standard Input. I noticed there was nothing written about this topic in the docs currently so I did some research and made this write-up.
<!-- Please write a summary of the change, such as which topic(s) and file(s) that you have edited or created. Please also include relevant motivation and context: -->

### Type of Change

<!--- Please delete or cross off the bullet point(s) that are irrelevant to this PR: -->

- Adding a new entry

### Checklist

<!-- Please check ALL the boxes: -->

- [ ] All writings are my own.
- [ ] My entry follows the Codecademy Docs style guide.
- [ ] My changes generate no new warnings.
- [ ] I have performed a self-review of my own writing and code.
- [ ] I have checked my entry and corrected any misspellings.
- [ ] I have made corresponding changes to the documentation if needed.
- [ ] I have confirmed my changes are not being pushed from my forked `main` branch.
- [ ] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [ ] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

<!--
Having trouble with the PR checker? Here are some common issues and resolutions:

- verify_formatting is failing
  - run `yarn format path/to/markdown/file.md` or `yarn format:all` and commit the results
- verify_lint is failing
  - same as above
  - if verify_lint is still failing, running `yarn lint` locally should let you know what needs to be changed by hand
- test is failing
  - ensure any new markdown files have a `Title` and `Description` defined in their metadata
  - ensure any new markdown files only contain alphanumerics and dashes in their file names and have the same name as their parent directory
  - if that looks ok, running `yarn test` locally should let you know what the issue is
-->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Codecademy/docs?pullRequest=2166) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Codecademy/docs?pullRequest=2166) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Codecademy/docs?pullRequest=2166) it.</sub>

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,130 @@
+---
+Title: 'ARGF and ARGV'
+Description: 'ARGF is a module in Ruby used for reading files passed as command line arguments or in standard input (STDIN) in a terminal. A Ruby script using ARGF may be thought of as implementing a subset of the functions of the Unix command `cat`.'
```
```suggestion
Description: 'ARGF is a module in Ruby used for reading files passed as command line arguments or in standard input (STDIN) in a terminal. A Ruby script using ARGF may be thought of as implementing a subset of the functions of the Unix command cat.'
```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,130 @@
+Subjects:
+  - 'Web Development'
+Tags:
+  - 'ARGF'
```
None of these tags are in the tags.md file. You need to select tags from that list. While you can add new tags in this PR to tags.md, "ARGF" and "ARGV" are too niche to be good tags for an entry. There is an "Input" tag and a "Files" tag that would be appropriate.

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,130 @@
+  - 'ARGF'
+  - 'ARGV'
+  - 'Standard Input'
+CatalogContent:
```
While we have two entries for CatalogContent, generally one is to one of the paid course paths, an appropriate one is `paths/front-end-engineer-career-path`

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,130 @@
+  - 'learn-rails'
+---
+
+ARFG and ARGV
```
Remove this table of contents, we do not use these in entries.

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,130 @@
+- [readlines](#readlines)
+
+
+ARGF is a module in Ruby used for reading files passed as command line arguments or in standard input (STDIN) in a terminal. A Ruby script using ARGF may be thought of as implementing a subset of the functions of the Unix command `cat`.
```
```suggestion
ARGF is a module in Ruby used for reading files passed as command line arguments or in standard input (STDIN) in a terminal. A Ruby script using ARGF may be thought of as implementing a subset of the functions of the Unix command `cat`.

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+
+
+ARGF is a module in Ruby used for reading files passed as command line arguments or in standard input (STDIN) in a terminal. A Ruby script using ARGF may be thought of as implementing a subset of the functions of the Unix command `cat`.
+```
````
````suggestion
```shell
````

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+echo "standard input" | cat
+# Output: standard input
+```
+The `cat` command can accept a filename or you can | pipe another command as standard input.
````
```suggestion

The `cat` command can accept a filename or another command can be piped (`|`) as standard input.
```
We generally do not use first or second person pronouns in Docs.

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+```
+The `cat` command can accept a filename or you can | pipe another command as standard input.
+
+### Reading files with Ruby
````
```suggestion
## ARGF - the F is for files
```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+# argf.rb
+puts ARFG.read
+```
+## ARGF - the F is for files
````
```suggestion

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+puts ARFG.read
+```
+## ARGF - the F is for files
+The ARGF module has a method #file that you can use to recognize what file it is currently handling. Create and save the file above and then enter this on the command line:
````
```suggestion

The ARGF module has a method #file that can be used to recognize what file it is currently handling. The following command line example uses the saved file above:

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+```
+## ARGF - the F is for files
+The ARGF module has a method #file that you can use to recognize what file it is currently handling. Create and save the file above and then enter this on the command line:
+```
````
````suggestion
```shell
````

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+# File 1 contents
+# File 2 contents
+```
+If no file is passed in, ARGF can also detect the standard input stream being piped in.
````
```suggestion

If no file is passed in, ARGF can also detect the standard input stream being piped in.

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+# File 2 contents
+```
+If no file is passed in, ARGF can also detect the standard input stream being piped in.
+```
````
````suggestion
```shell
````

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+# Output:
+# STDIN
+```
+A combination of files and standard input is accepted too.
````
```suggestion

A combination of files and standard input is accepted too.

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+# STDIN
+```
+A combination of files and standard input is accepted too.
+```
````
````suggestion
```shell
````

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+```
+
+## ARGV - the V is for variable
+Behind the scenes, inside the file there is a Ruby array called ARGV which contains the names of all arguments passed to the file at the command line or through STDIN.
````
```suggestion

Behind the scenes, inside the file there is a Ruby array called ARGV which contains the names of all arguments passed to the file at the command line or through STDIN.

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+# argf_inspect_argv.rb
+puts ARGV.inspect
+```
+The ARGV array doesn't care whether the arguments are valid filenames or not.
````
```suggestion

The ARGV array doesn't care whether the arguments are valid filenames or not.

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+puts ARGV.inspect
+```
+The ARGV array doesn't care whether the arguments are valid filenames or not.
+```
````
````suggestion
```shell
````

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+# Output: ["one", "two"]
+```
+
+## shift
````
```suggestion
## `shift`

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,130 @@
+ARGF is using ARGV to store the names of the files passed into it at the command line.
+Files that are processed by ARGF.read are shifted off the ARGV array
+
+However, as is often the case with terminal commands, flags like --verbose or --all are often used.
```
```suggestion
However, as is often the case with terminal commands, flags like `--verbose` or `--all` are often used.
```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+
+However, as is often the case with terminal commands, flags like --verbose or --all are often used.
+What happens in a case like this?
+```
````
````suggestion

```shell
````

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+ruby argf_inspect_argv.rb --all one two
+# Output: ["--all", "one", "two"]
+```
+We should use the #shift method of ARGV
````
```suggestion

The #shift method of ARGV should be used:

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+ARGV.shift
+puts ARGV.inspect
+```
+```
````
````suggestion

```shell
````

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+["one", "two"]
+```
+
+## readlines
````
```suggestion
## `readlines`

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+```
+
+## readlines
+The method of ARGF #readlines created an array of strings of each line of each file passed to the Ruby script.
````
```suggestion

The method of ARGF #readlines creates an array of strings of each line of each file passed to the Ruby script.

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+puts lines
+puts lines.inspect
+```
+```
````
````suggestion

```shell
````

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+# Contents of file2
+# ["Contents of file1\n", "Contents of file2\n"]
+```
+#readlines will not accept command line arguments that are not valid filenames.
````
```suggestion

`#readlines` will not accept command line arguments that are not valid filenames.

```

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,130 @@
+ruby argv_readlines.rb one two
+=> argv_readlines.rb:1:in `readlines': No such file or directory @ rb_sysopen - one (Errno::ENOENT) from argv_readlines.rb:1:in `<main>'
+```
+### References
````
Remove the references section as we generally don't use these in Docs.

SSwiniarski reviewed (commented):
@tmravic I did a first review and left a few comments. There may be more once these are addressed as these were mostly formatting issues.

tmravic reviewed (commented):
@SSwiniarski
The comments you left previously should be mostly resolved. Please have another look when you can.

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,131 @@
+  - 'Files'
+CatalogContent:
+  - 'learn-ruby'
+  - 'front-end-engineer-career-path'
```
```suggestion
  - 'paths/front-end-engineer-career-path'
```

SSwiniarski reviewed (changes requested):
@tmravic A few more comments.

tmravic commented:
@SSwiniarski
I fixed the further issues you brought up/

Also, I accidentally rebased main onto my branch instead of merging it.
So I did an interactive rebase and dropped the commits from main.
Then I did a force push, then I merged main and pushed that.
Hope I didn't break anything but I only see my own changes on this PR.
I'm still learning how to use `git rebase` so I apologize for the mistake.

SSwiniarski commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,130 @@
+  - 'learn-ruby'
+  - 'paths/front-end-engineer-career-path'
+---
+
```
```suggestion
```
Remove line break.

SSwiniarski reviewed (approved):
@tmravic Aside from one more line break, looks good for a second review.

yangc95 commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,132 @@
+---
+Title: 'ARGF and ARGV'
```
```suggestion
Title: 'ARGF'
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,132 @@
+# Output: standard input
+```
+
+The cat command can accept a filename or you can | pipe another command as standard input.
````
```suggestion
The `cat` command can accept a filename or it can pipe (`|`) another command as standard input.
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,132 @@
+
+The cat command can accept a filename or you can | pipe another command as standard input.
+
+## ARGF - the F is for files
```
```suggestion
## ARGF - The F Is for Files
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,132 @@
+
+```ruby
+# argf.rb
+puts ARFG.read
````
```suggestion
puts ARGF.read
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,132 @@
+puts ARFG.read
+```
+
+The ARGF module has a method #file that you can use to recognize what file it is currently handling. Type this on the command line:
````
```suggestion
The ARGF module has a method `.file` that can be used to recognize what file it is currently handling. Type this on the command line:
```

Is this method being used below?

tmravic replied:
I have replaced this with a description of the `.read` method instead.

yangc95 commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,132 @@
+# STDIN
+```
+
+## ARGV - the V is for variable
````
```suggestion
## ARGV - The V Is for Variable
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,132 @@
+
+## ARGV - the V is for variable
+
+Behind the scenes, inside the file there is a Ruby array called ARGV which contains the names of all arguments passed to the file at the command line or through STDIN.
```
```suggestion
Under the hood and inside the file, there is a Ruby array termed `ARGV` which contains the names of all arguments passed to the file at the command line or through STDIN.
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,132 @@
+puts ARGV.inspect
+```
+
+The ARGV array doesn't care whether the arguments are valid filenames or not.
````
```suggestion
The `ARGV` array disregards whether the arguments are valid filenames or not.
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,132 @@
+## `shift`
+
+ARGF is using ARGV to store the names of the files passed into it at the command line.
+Files that are processed by ARGF.read are shifted off the ARGV array
```
```suggestion

Files that are processed by `ARGF.read` are shifted off the `ARGV` array
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,132 @@
+ARGF is using ARGV to store the names of the files passed into it at the command line.
+Files that are processed by ARGF.read are shifted off the ARGV array
+
+However, as is often the case with terminal commands, flags like `--verbose` or `--all` are often used.
+What happens in a case like this?
```
```suggestion
However, as is often the case with terminal commands, flags like `--verbose` or `--all` are often used.
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,132 @@
+
+## `shift`
+
+ARGF is using ARGV to store the names of the files passed into it at the command line.
```
This line along with the following lines don't introduce the `shift` method very well. It's a bit jumbled. It would be best to go straight into the method by explaining what it does and is used for. Other extraneous info can be placed in a note which would be formatted like:

> **Note:** Some note here

yangc95 commented on content/ruby/concepts/argf/argf.md:
````diff
@@ -0,0 +1,132 @@
+# ["Contents of file1\n", "Contents of file2\n"]
+```
+
+`#readlines` will not accept command line arguments that are not valid filenames.
````
```suggestion
`.readlines` will not accept command line arguments that are not valid filenames.
```

yangc95 commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,132 @@
+
+## `readlines`
+
+The method of ARGF #readlines created an array of strings of each line of each file passed to the Ruby script.
```
This is good! It explains what the method does. It would be good to also include an example use-case like "It's commonly used for...."

yangc95 reviewed (commented):
@tmravic I conducted a second review! Please let me know if you have any questions or concerns.

After resolving the comments/suggestions, please tag me and I can take another look.

tmravic commented on content/ruby/concepts/argf/argf.md:
```diff
@@ -0,0 +1,132 @@
+  - 'paths/front-end-engineer-career-path'
+---
+
+ARGF is a module in Ruby used for reading files passed as command line arguments or in standard input (STDIN) in a terminal. A Ruby script using ARGF may be thought of as implementing a subset of the functions of the Unix command cat.
```
The previous reviewer told me to leave off the backticks here. What should I do?
https://github.com/Codecademy/docs/pull/2166#discussion_r1193143433

tmravic replied:
@yangc95 I'm not sure if you saw my previous tag, please take another look at let me know what else I can fix.

yangc95 replied:
@tmravic that was for the metadata's description. Having the backticks in the actual content (introduction section) is good

tmravic reviewed (commented):
@yangc95 
I have made most of the the changes you suggested, please take another look.

yangc95 commented:
🥳 @tmravic live link: https://www.codecademy.com/resources/docs/ruby/argf

<img width="808" alt="Screenshot 2023-05-30 at 1 13 37 PM" src="https://github.com/Codecademy/docs/assets/72277593/cb0cb4bb-c1be-482c-89eb-76b4dbd6feff">

cc: @SSwiniarski
